k2/python/csrc/torch/index_select.cu

```diff
@@ -0,0 +1,143 @@
+/**
+ * @brief Index select for k2.
+ *
+ * Unlike torch.index_select, when an entry is -1, it sets
+ * the destination entry to 0.
+ *
+ * @copyright
+ * Copyright (c)  2020  Mobvoi Inc.        (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include "k2/csrc/context.h"
+#include "k2/python/csrc/torch/index_select.h"
+#include "k2/python/csrc/torch/torch_util.h"
+#include "torch/extension.h"
+
+namespace k2 {
+
+template <typename T>
+static torch::Tensor IndexSelect1D(torch::Tensor src, torch::Tensor index) {
+  K2_CHECK_EQ(src.dim(), 1) << "Expected dim: 1. Given: " << src.dim();
+  K2_CHECK_EQ(src.scalar_type(), ToScalarType<T>::value);
+  K2_CHECK(src.is_contiguous());
+
+  K2_CHECK(index.is_contiguous());
+  K2_CHECK_EQ(index.dim(), 1);
+  K2_CHECK_EQ(index.scalar_type(), ToScalarType<int32_t>::value);
+
+  ContextPtr context;
+  if (src.device().type() == torch::kCPU) {
+    context = GetCpuContext();
+  } else {
+    K2_CHECK(src.is_cuda());
+    context = GetCudaContext(src.device().index());
+  }
+
+  const T *src_data = src.data_ptr<T>();
+  int32_t src_numel = src.numel();
+  const int32_t *index_data = index.data_ptr<int32_t>();
+
+  torch::Tensor ans = torch::empty(index.sizes(), src.options());
+  T *ans_data = ans.data_ptr<T>();
+
+  auto lambda = [=] __host__ __device__(int32_t i) {
+    int32_t src_i = index_data[i];
+    if (src_i != -1) {
+      K2_DCHECK_GE(src_i, 0);
+      K2_DCHECK_LT(src_i, src_numel);
+
+      ans_data[i] = src_data[src_i];
+    } else {
+      ans_data[i] = 0;
+    }
+  };
+  Eval(context, index.numel(), lambda);
+  return ans;
+}
+
+static torch::Tensor IndexSelect1DBackward(torch::Tensor src,
```

Comment:
Incidentally, the core part of this function is the same as torch's index_add_, and we might eventually want to make that callable independently.  Just something to think about; no change required.

Reply:
Added `index_add_`.

Reply:
Do this neeed to be different from index_add, i.e. can't you just call index_add?

I don't think index_add_ is the right name if it's not a member function (?)   Should be index_add?

Reply:
> Do this need to be different from index_add, i.e. can't you just call index_add?

Yes, we still need it. 

`index_add_` does not handle the case where an index may be -1.
If we do not need to consider that indexes may be -1, then `IndexSelect1DBackward` can be
actually removed. We can call `index_add_` in Python for the backpropagation, which is
the case for `intersect_dense_pruned`.

Reply:
Fixed by removing `IndexSelect1DBackward`.